kernel/events.c

```diff
@@ -293,16 +305,23 @@ static uint32_t k_event_wait_internal(struct k_event *event, uint32_t events,
 	SYS_PORT_TRACING_OBJ_FUNC_BLOCKING(k_event, wait, event, events,
 					   options, timeout);
 
-	if (z_pend_curr(&event->lock, key, &event->wait_q, timeout) == 0) {
-		/* Retrieve the set of events that woke the thread */
-		rv = thread->events;
+	for (k_timepoint_t end = sys_timepoint_calc(timeout); !K_TIMEOUT_EQ(timeout, K_NO_WAIT);
+	     timeout = sys_timepoint_timeout(end)) {
+
```

Comment:
I am assuming that it is meant for some sort or retry logic, which experience has taught me is tricky. Upon the 2nd pass through this loop the event object's spinlock is going to be unlocked, which will violate an assumption of z_pend_curr().

Reply:
Yeah, I think there is a minor race condition there. Is there any version of pend that retains the lock?

Reply:
No. But one thing I am trying to figure out is why have the retry condition at all?

Reply:
It's probably not needed anymore. Originally, I thought it might fix something unusual I saw, but that was fixed with a modification to the post codepath

Reply:
Fixed!